File: cesta/SergioDeIscar-cesta-java/SergioDeIscar-cesta-java/src/main/java/org/example/models/ItemFactura.java
package org.example.models;

public final class ItemFactura {
    private int cantidad;
    private Producto producto;
    private final float precioUnitario;

    public ItemFactura(Producto producto, int cantidad) {
        if (producto == null)
            throw new IllegalArgumentException("El producto no puede ser nulo");
        if (cantidad <= 0)
            throw new IllegalArgumentException("La cantidad debe ser mayor a 0");
        setCantidad(cantidad);
        setProducto(producto);
        precioUnitario = producto.getPrecio();
    }

    //region Getters
    public int getCantidad() {
        return cantidad;
    }
    public Producto getProducto() {
        return producto;
    }
    public float getPrecioUnitario() {
        return precioUnitario;
    }
    //endregion
    //region Setters
    public void setCantidad(int cantidad) {
        if (cantidad <= 0)
            throw new IllegalArgumentException("La cantidad debe ser mayor a 0");
        this.cantidad = cantidad;
    }
    public void setProducto(Producto producto) {
        if (producto == null)
            throw new IllegalArgumentException("El producto no puede ser nulo");
        this.producto = producto;
    }
    //endregion

    public float calcularImporte() {
        return this.cantidad * precioUnitario;
    }

    @Override
    public String toString() {
        return "ItemFactura -> Producto -> " + this.producto + ", Cantidad -> " + this.cantidad + ", Precio Unitario -> " + precioUnitario;
    }
}

File: cesta/SergioDeIscar-cesta-kotlin/SergioDeIscar-cesta-kotlin/src/main/kotlin/models/ItemFactura.kt
package factura.models

class ItemFactura (val item: Producto){
    var cantidad: Int = 1
        set(value) {
            require(field > 0 && field <= item.stock)
            field = value
        }
    val precioUnitario: Float = item.GetPrecio()

    constructor(item: Producto, cantidad: Int): this(item){
        this.cantidad = cantidad
    }

    override fun toString(): String {
        return "ItemFactura -> Producto -> { $item }, Cantidad -> $cantidad, Precio Unitario -> $precioUnitario"
    }
}